### PR TITLE
Remove Makefile, put everything in npm scripts

### PR DESCRIPTION
**Pull request by targos**
Fixes: #559

### CODE REVIEW

**Comment by targos** [comment]
I just don't know what to do with the benchmarks

File: package.json [hunk 1]
```diff
@@ -4,7 +4,10 @@
   "description": "Koa web app framework",
   "main": "lib/application.js",
   "scripts": {
-    "test": "make test"
+    "lint": "eslint lib test",
+    "test": "NODE_ENV=test mocha --require should --require should-http test/application/* test/context/* test/request/* test/response/*",
+    "test-cov": "NODE_ENV=test istanbul cover ./node_modules/.bin/_mocha -- -u exports --require should --require should-http test/application/* test/context/* test/request/* test/response/* --bail",
```
Comment (stojanovic) [inline package.json]: `_mocha` is already in the path in scripts, no need for `.node_modules/.bin/_mocha`, this should work fine:
```
"test-cov": "NODE_ENV=test istanbul cover _mocha -- -u exports --require should --require should-http test/application/* test/context/* test/request/* test/response/* --bail",
```

**Comment by juliangruber** [comment]
-1, make is a lot faster than pm

File: package.json (hunk 1 above)
Comment (targos) [inline package.json]: you're right, I changed it.

**Comment by targos** [comment]
Timings on my side:
```
make test > /dev/null  0.65s user 0.07s system 106% cpu 0.675 total
npm test > /dev/null  1.12s user 0.13s system 100% cpu 1.240 total
```
I personally don't care, 1-2 seconds to run tests is fast enough.
Seems that it doesn't change anything on travis (always around 50s)

**Comment by stojanovic** [comment]
@juliangruber you are right, it is faster, also we still have Makefile for benchmarks.
But people on windows needs to run tests and coverage manually. Maybe we can keep both, and use Make on travis, but who ever wants can use one or another.
It's a bit redundant of course.

**Comment by felixfbecker** [comment]
This is what I get when I run with make in an Ubuntu VM:
```
/media/sf_git/koa/node_modules/.bin/_mocha:4
case `uname` in
^^^^

SyntaxError: Unexpected token case
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:414:25)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Function.Module.runMain (module.js:467:10)
    at startup (node.js:134:18)
    at node.js:961:3
```
But why do I have to go into an Ubuntu VM just so I can run a make file that then calls js files that could be easily put into npm scripts?  
I don't quite get why `make` should be significantly faster - in the end those two are only two ways of launching js files.

Btw in Windows I think you can only set env variables with `set` command.

**Comment by felixfbecker** [comment]
`env NODE_ENV=test mocha` should work cross-platform if we really need the NODE_ENV, assuming you have git installed and in path on Windows.

**Comment by felixfbecker** [comment]
But imo using stuff like NODE_ENV is bad practice anyways (see http://12factor.net/config). I looked at the test code and actually never uses it, so why does need to be there?
The NODE_ENV-dependent error handling is annotated as deprecated.
There is one test in `test/application/toJSON.js` that assumes the env variable of `test`, when it really should do it like `test/application/index.js:50` and set the env beforehand and reset after.

**Comment by felixfbecker** [comment]
For benchmarks just keep the Makefile in the benchmarks subdir. Oh and btw instead of 
`mocha --require should --require should-http test/application/* test/context/* test/request/* test/response/*`
we can just write
`mocha --require should --require should-http --recursive` (mocha looks in test directory by default)
or
`mocha --require should --require should-http test/**/*.js`
But now enough spam for this thread :D

**Comment by yanickrochon** [comment]
I agree with @felixfbecker about `NODE_ENV` being misused. We can change `process.env` when testing (self contained tests) and relying on external settings is bad maintenance practice.

**Comment by jonathanong** [comment]
don't have an opinion here...

but if we move it to npm, we should move the config to an `mocha.opts` file or something so the commands aren't 2000 characters long

**Comment by felixfbecker** [comment]
Like I said, you don't need to specify the paths in mocha. It could be as short as
```sh
mocha -r should -r should-http --recursive
```
will look inside test folder and all subfolders.
Also, I think we agreed on removing NODE_ENV setting (would require one change in one of the test as mentioned).

**Comment by targos** [comment]
@felixfbecker I updated according to your comments. PTAL

**Comment by felixfbecker** [comment]
looks great :+1:

File: package.json [hunk 2]
```diff
@@ -4,7 +4,10 @@
   "description": "Koa web app framework",
   "main": "lib/application.js",
   "scripts": {
-    "test": "make test"
```
Comment (jonathanong) [inline package.json]: we need a `npm run bench` task too then

**Comment by jonathanong** [comment]
@koajs/owners thoughts?

**Comment by tj** [comment]
I find it really weird that running scripts is in npm's domain, but no major complaints here

**Comment by yanickrochon** [comment]
@tj I presume this is because npm *can* execute pre-install and post-install scripts... so the feature is already there. At some point, I guess it just became standard to run tests and stuff with npm. Personally, I think it's somewhat within it's domain, as npm provides tools (i.e. running scripts) to ensure some package quality. Anyhow, it's much easier to write script instructions in `package.json` than creating a Makefile... at least to people who never wrote such files before :)

**Comment by travisjeffery** [comment]
-1 here too, make is faster and more flexible and is currently the one place you need to look to find the testing/benchmarking stuff which makes it more approachable.

**Comment by felixfbecker** [comment]
I don't think it is more approachable. Every Node developer knows npm scripts. They look at a package.json and immidietly know what it does. They expect to do `npm test` to run the tests after they forked because that works across all modules from npm. A makefile is a complicated piece of file that only works on specific platforms and does nothing what npm scripts cannot do. If you need more complicated build scripts, write a js file and call it in your npm script.

**Comment by juliangruber** [comment]
The package.json is still using the bench script, can someone add that?

**Comment by tj** [comment]
It's just another thing that npm does poorly, but seeing as windows people seemed to have never come up with a reasonable Make implementation (no reason they couldn't) I'm cool with it

**Comment by yanickrochon** [comment]
Using `make` shaves a mere second or two when running tests over `npm`. As for benchmark, they are totally relative and being faster with a Makefile means nothing at all since they cannot be compared between computers anyhow. So, really, it all comes down to a preference.

That being said, as @felixfbecker said, every Node programmer expect packages to provide `npm test` to run the test suites. And since either solutions are standards and really no big deal in terms of maintenance, why don't we have both of them? Even as a Linux user, I don't personally use `make` (force of habit, I guess) and Windows support is inevitable. I cannot conceive writing apps in Windows that will be running on Linux... but I presume some people cannot live without a full blown IDE like VS, but having an implemented standard (i.e. `npm test`) be available is irrelevant to the OS. It feels like the old war between using spaces or tabs... except that, now, we can have both :)

**Comment by dead-horse** [comment]
please change the target branch to `v2.x`.

**Comment by felixfbecker** [comment]
Why? This is just dev stuff.

**Comment by tejasmanohar** [comment]
@targos @felixfbecker I think it's fine to send a PR for this to `master` as long as it doesn't bring all the `v2.x` commits over (since the branches switched around). If that's merged, should probably send a PR to `v2.x` as well.

**Comment by targos** [comment]
I'm fine if someone else wants to take care of this. I am on holiday for the next two weeks.

**Comment by jonathanong** [comment]
@targos i'm going to close this because i don't think we've gotten a consensus and the PR is out of date. sorry :( thanks for PR though

**Comment by markelog** [comment]
Why an [issue](https://github.com/koajs/koa/issues/559) for this is still open?

**Comment by PlasmaPower** [comment]
@markelog because it's still a pain for windows users to run tests

**Comment by markelog** [comment]
So this was closed only because pr had some conflicts?

**Comment by PlasmaPower** [comment]
@markelog It was left to rot because of a number of reasons, mainly that it resulted in slower performance and that there aren't any Koa contributers on Windows that we know of.

**Comment by felixfbecker** [comment]
@PlasmaPower It's a chicken-or-egg problem ;)

**Comment by markelog** [comment]
@PlasmaPower your comments contradict to each other :/, if pr for this is not acceptable, issue should be closed

**Comment by PlasmaPower** [comment]
@markelog There might be another solution, IDK I haven't looked into it. Maybe we could not use the Makefile only for Windows users?

@felixfbecker Yep, somewhat, but you can run the tests with cygwin/mingw anyway.

**Comment by jeffijoe** [comment]
@PlasmaPower we probably won't have any Windows code contribs if we can't support it. ;)

**Comment by PlasmaPower** [comment]
@jeffijoe As I've commented above, most Windows devs are running Cygwin/Mysys anyway and therefore have access to Make. However, if we have enough of them it would be helpful to switch to native NPM scripts.

**Comment by jeffijoe** [comment]
I'm running MINGW which ships with Git and I don't have make.

**Comment by PlasmaPower** [comment]
@jeffijoe Install MYSYS.

**Pull request closed without merging**